raiden_contracts/data/source/raiden/TokenNetwork.sol

```diff
@@ -479,7 +506,74 @@ contract TokenNetwork is Utils {
+    /// operations can be done on it.
+    /// An important constraint is that this function checks that all tokens
+    /// in this channel are withdrawn. This means that the channel can *not*
+    /// have any outstanding locked transfers.
+    /// @param channel_identifier Identifier for the channel on which this
+    /// operation takes place
+    /// @param data1 Withdraw data of the participant initiating the cooperative settlement
+    /// @param data2 Withdraw data of the second participant
+    function cooperativeSettle(
+        uint256 channel_identifier,
+        WithdrawInput memory data1,
+        WithdrawInput memory data2
+    )
+        external
+        isOpen(channel_identifier)
+    {
+        uint256 total_deposit;
+        bytes32 pair_hash;
+
+        // Validate that authenticated partners and the channel identifier match
+        require(channel_identifier == getChannelIdentifier(data1.participant, data2.participant));
+        Channel storage channel = channels[channel_identifier];
+
+        Participant storage participant1_state = channel.participants[data1.participant];
+        Participant storage participant2_state = channel.participants[data2.participant];
+        total_deposit = participant1_state.deposit + participant2_state.deposit;
 
+        // The sum of the provided balances must be equal to the total
+        // available deposit. This also implies that no locks must exist in the channel
+        // when this is called, as otherwise the withdrawable amount would be smaller
+        // than required.
+        require((data1.total_withdraw + data2.total_withdraw) == total_deposit, "CS: incomplete amounts");
+        // Overflow check for the balances addition from the above check.
+        // This overflow should never happen if the token.transfer function is implemented
+        // correctly. We do not control the token implementation, therefore we add this
+        // check for safety.
+        require(data1.total_withdraw <= data1.total_withdraw + data2.total_withdraw);
+
+        if (data1.total_withdraw > 0) {
+            this.setTotalWithdraw(
+                channel_identifier,
+                data1.participant,
+                data1.total_withdraw,
+                data1.expiration_block,
+                data1.participant_signature,
+                data1.partner_signature
+            );
+        }
+        if (data2.total_withdraw > 0) {
+            this.setTotalWithdraw(
+                channel_identifier,
+                data2.participant,
+                data2.total_withdraw,
+                data2.expiration_block,
+                data2.participant_signature,
+                data2.partner_signature
+            );
+        }
+        removeChannelData(channel, channel_identifier, data1.participant, data2.participant);
+
+        // Emit channel lifecycle events
+        emit ChannelClosed(channel_identifier, data1.participant, 0, 0);
```

Comment:
I'm not sure if we should emit this event, since the channel will never be in a closed state. I think I prefer to have it to make both settlement approaches more similar, but it does not match the concept fully (the nonce and balance_hash do not really apply) and removing it would also be slightly cheaper.

Reply:
The intention was to keep the same channel lifecycle as before. In other words, I am not sure if the PFS/explorer/... rely on the close event in some sense.

But I will check on this again.

Reply:
Seems like we need some adjustments in either case (with or without the `ChannelClosed` event):

- The PFS currently relies on the `ChannelClosed` event, and doesn't even handle the `ChannelSettled` event:
https://github.com/raiden-network/raiden-services/blob/37c6f4816e2a6a084bd4e7e9c0eb58b501870b66/src/pathfinding_service/service.py#L228-L229
  - However, adding the same handling for `ChannelSettled` events shouldn't be difficult
- The MS triggers a `ActionMonitoringTriggeredEvent` (https://github.com/raiden-network/raiden-services/blob/37c6f4816e2a6a084bd4e7e9c0eb58b501870b66/src/monitoring_service/handlers.py#L124) when a channel is closed. The event handler (https://github.com/raiden-network/raiden-services/blob/37c6f4816e2a6a084bd4e7e9c0eb58b501870b66/src/monitoring_service/handlers.py#L435) for this event doesn't currently checked if the channel in question is settled already.


tldr: We need to adjust the services anyways, so we can decide freely on which approach makes more logical sense.

Reply:
We agreed to not emit the `ChannelClosed` event.